runtime/src/accounts_db.rs

```diff
@@ -3651,14 +3663,219 @@ impl AccountsDB {
             let success = hash == hash_other
                 && total_lamports == total_lamports_other
                 && total_lamports == expected_capitalization.unwrap_or(total_lamports);
-            assert!(success, "update_accounts_hash_with_index_option mismatch. hashes: {}, {}; lamports: {}, {}; expected lamports: {:?}, using index: {}, slot: {}", hash, hash_other, total_lamports, total_lamports_other, expected_capitalization, use_index, slot);
+            let msg = format!("update_accounts_hash_with_index_option mismatch. hashes: {}, {}; lamports: {}, {}; expected lamports: {:?}, using index: {}, slot: {}", hash, hash_other, total_lamports, total_lamports_other, expected_capitalization, use_index, slot);
+
+            if !success {
+                error!("prior to panic, trying to identify differences: {}", msg);
+                self.debug_compare_hash_calculation_details(
+                    slot,
+                    ancestors,
+                    simple_capitalization_enabled,
+                );
+            }
+
+            assert!(success, msg);
         }
+
         let mut bank_hashes = self.bank_hashes.write().unwrap();
         let mut bank_hash_info = bank_hashes.get_mut(&slot).unwrap();
         bank_hash_info.snapshot_hash = hash;
         (hash, total_lamports)
     }
 
+    // only called prior to a panic.
+    // The goal is to dump out account information to help diagnose the source of differences.
+    fn debug_compare_hash_calculation_details(
+        &self,
+        slot: Slot,
+        ancestors: &Ancestors,
+        simple_capitalization_enabled: bool,
+    ) -> bool {
+        // this code path will only execute when debug_verify is true
+        // this code is cobbled together from the normal execution path
+        let storages = self.get_snapshot_storages(slot);
+        let scan_and_hash = || {
+            let mut at_least_one_mismatch = false;
+            let mut stats = HashStats::default();
+            // clipped and modified from calculate_accounts_hash_without_index
```

Comment:
It's a lot of copied code, seems like a maintenance burden for just a debug function. Can we go the way have having two paths which compute the intermediates in both ways which is shared with the normal code and then doing like a post-processing on both of them to determine the diff?